File: src/DashConnect.Core/Update/UpdateChecker.cs
using System.Net.Http;
using System.Net.Http.Headers;
using System.Security.Cryptography;
using System.Text.Json;
using System.Text.RegularExpressions;
using DashConnect.Core.Logging;

namespace DashConnect.Core.Update;

/// <summary><paramref name="Sha256Url"/> points at a sibling <c>&lt;asset&gt;.sha256</c> file whose
/// first 64-hex token is the expected digest of the installer; null when the release didn't publish
/// one (older releases).</summary>
public sealed record UpdateInfo(string Version, string DownloadUrl, string Notes, string? Sha256Url = null);

/// <summary>
/// Checks GitHub Releases for a newer version and downloads the installer. The current version is a
/// single source of truth used both here and by the MSI/product build.
/// </summary>
public static class UpdateChecker
{
    public const string CurrentVersion = "1.1.15";
    public const string Owner = "dutow20162007-create";
    public const string Repo = "DashConnect";

    public static string ReleasesPage => $"https://github.com/{Owner}/{Repo}/releases/latest";

    /// <summary>Returns update info if a newer release exists, otherwise null (never throws). Tries
    /// the rich GitHub API first, then a github.com redirect (more reachable on RU ISPs).</summary>
    public static async Task<UpdateInfo?> CheckAsync(CancellationToken ct = default)
    {
        var latest = await LatestViaApiAsync(ct) ?? await LatestViaRedirectAsync(ct);
        if (latest is null) return null;
        return IsNewer(latest.Version, CurrentVersion) ? latest : null;
    }

    /// <summary>Latest release via api.github.com (rich: real asset URL + notes). Null on failure.</summary>
    private static async Task<UpdateInfo?> LatestViaApiAsync(CancellationToken ct)
    {
        try
        {
            using var http = NewClient(TimeSpan.FromSeconds(15));
            http.DefaultRequestHeaders.Accept.Add(new MediaTypeWithQualityHeaderValue("application/vnd.github+json"));
            var json = await http.GetStringAsync(
                $"https://api.github.com/repos/{Owner}/{Repo}/releases/latest", ct);

            using var doc = JsonDocument.Parse(json);
            var root = doc.RootElement;
            var tag = (root.TryGetProperty("tag_name", out var t) ? t.GetString() : null)?.TrimStart('v', 'V') ?? "";
            if (string.IsNullOrWhiteSpace(tag)) return null;

            var notes = root.TryGetProperty("body", out var b) ? b.GetString() ?? "" : "";
            string? msi = null, sha = null;
            if (root.TryGetProperty("assets", out var assets) && assets.ValueKind == JsonValueKind.Array)
                foreach (var a in assets.EnumerateArray())
                {
                    var name = a.TryGetProperty("name", out var n) ? n.GetString() ?? "" : "";
                    if (name.EndsWith(".msi", StringComparison.OrdinalIgnoreCase))
                        msi = a.GetProperty("browser_download_url").GetString();
                    else if (name.EndsWith(".msi.sha256", StringComparison.OrdinalIgnoreCase))
                        sha = a.GetProperty("browser_download_url").GetString();
                }
            return new UpdateInfo(tag, msi ?? ReleasesPage, notes, sha);
        }
        catch (Exception ex)
        {
            Log.Debug("update", $"api.github.com недоступен: {ex.Message}");
            return null;
        }
    }

    /// <summary>Fallback: read the tag from the github.com/…/releases/latest redirect and build the
    /// asset URL from the known naming pattern. api.github.com is often blocked in RU; github.com
    /// (especially with the DPI bypass on) is far more reachable.</summary>
    private static async Task<UpdateInfo?> LatestViaRedirectAsync(CancellationToken ct)
    {
        try
        {
            using var handler = new HttpClientHandler { AllowAutoRedirect = false };
            using var http = new HttpClient(handler) { Timeout = TimeSpan.FromSeconds(15) };
            http.DefaultRequestHeaders.UserAgent.Add(new ProductInfoHeaderValue("DashConnect", CurrentVersion));
            var resp = await http.GetAsync($"https://github.com/{Owner}/{Repo}/releases/latest", ct);
            var location = resp.Headers.Location?.ToString() ?? "";
            var m = Regex.Match(location, @"/tag/v?([0-9][0-9.]*)");
            if (!m.Success) return null;
            var tag = m.Groups[1].Value;
            var msi = $"https://github.com/{Owner}/{Repo}/releases/download/v{tag}/DashConnect-{tag}.msi";
            return new UpdateInfo(tag, msi, "", msi + ".sha256");
        }
        catch (Exception ex)
        {
            Log.Debug("update", $"github.com редирект недоступен: {ex.Message}");
            return null;
        }
    }

    /// <summary>Downloads the installer to a temp file and returns its path (null on failure). When the
    /// release published a <c>.sha256</c>, the file is verified against it and a MISMATCH fails the
    /// download (returns null) — we never hand a tampered/corrupt installer to msiexec.</summary>
    public static async Task<string?> DownloadAsync(
        string url, IProgress<double>? progress = null, string? sha256Url = null, CancellationToken ct = default)
    {
        try
        {
            var dest = Path.Combine(Path.GetTempPath(), $"DashConnect-{Guid.NewGuid():N}.msi");
            using var http = NewClient(TimeSpan.FromMinutes(10));
            using var resp = await http.GetAsync(url, HttpCompletionOption.ResponseHeadersRead, ct);
            resp.EnsureSuccessStatusCode();
            var total = resp.Content.Headers.ContentLength ?? -1L;

            await using (var src = await resp.Content.ReadAsStreamAsync(ct))
            await using (var dst = File.Create(dest))
            {
                var buffer = new byte[81920];
                long read = 0;
                int n;
                while ((n = await src.ReadAsync(buffer, ct)) > 0)
                {
                    await dst.WriteAsync(buffer.AsMemory(0, n), ct);
                    read += n;
                    if (total > 0) progress?.Report((double)read / total);
                }
            }

            if (!string.IsNullOrWhiteSpace(sha256Url) && !await VerifyChecksumAsync(dest, sha256Url, ct))
            {
                Log.Error("update", "контрольная сумма обновления не совпала — установка отменена");
                try { File.Delete(dest); } catch { }
                return null;
            }
            return dest;
        }
        catch (Exception ex)
        {
            Log.Warn("update", $"скачивание обновления не удалось: {ex.Message}");
            return null;
        }
    }

    /// <summary>Downloads the expected digest and compares it to the file's actual SHA-256. On any
    /// error fetching the checksum it returns true (fail-open — an older release without a .sha256
    /// must still be installable); only a real, confirmed mismatch returns false.</summary>
    private static async Task<bool> VerifyChecksumAsync(string file, string sha256Url, CancellationToken ct)
    {
        string expected;
        try
        {
            using var http = NewClient(TimeSpan.FromSeconds(30));
            var text = await http.GetStringAsync(sha256Url, ct);
            var m = Regex.Match(text, "[0-9a-fA-F]{64}");
            if (!m.Success) { Log.Debug("update", "в .sha256 нет валидного хеша — пропускаю проверку"); return true; }
            expected = m.Value.ToLowerInvariant();
        }
        catch (Exception ex) { Log.Debug("update", $".sha256 недоступен ({ex.Message}) — пропускаю проверку"); return true; }

        await using var fs = File.OpenRead(file);
        var hash = Convert.ToHexString(await SHA256.HashDataAsync(fs, ct)).ToLowerInvariant();
        if (hash == expected) { Log.Info("update", "контрольная сумма обновления подтверждена (SHA-256)"); return true; }
        Log.Error("update", $"SHA-256 не совпал: ожидалось {expected}, получено {hash}");
        return false;
    }

    private static HttpClient NewClient(TimeSpan timeout)
    {
        var http = new HttpClient { Timeout = timeout };
        http.DefaultRequestHeaders.UserAgent.Add(new ProductInfoHeaderValue("DashConnect", CurrentVersion));
        return http;
    }

    /// <summary>Numeric dotted-version comparison (1.2.0 &gt; 1.1.9); non-numeric parts count as 0.</summary>
    private static bool IsNewer(string remote, string local)
    {
        static int[] Parse(string v) =>
            v.Split('.', '-', '+').Select(p => int.TryParse(p, out var num) ? num : 0).ToArray();
        var r = Parse(remote);
        var l = Parse(local);
        for (int i = 0; i < Math.Max(r.Length, l.Length); i++)
        {
            int rv = i < r.Length ? r[i] : 0, lv = i < l.Length ? l[i] : 0;
            if (rv != lv) return rv > lv;
        }
        return false;
    }
}

File: mobile/app/src/main/java/io/github/dovecoteescapee/byedpi/ui/UpdateChecker.kt
package io.github.dovecoteescapee.byedpi.ui

import android.content.Context
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.json.JSONObject
import java.io.File
import java.net.HttpURLConnection
import java.net.URL
import java.security.MessageDigest

/** [sha256Url] points at a sibling `<apk>.sha256` (null on older releases without one). */
data class UpdateInfo(val version: String, val apkUrl: String, val notes: String, val sha256Url: String? = null)

/**
 * The phone counterpart of the PC UpdateChecker: looks up the newest GitHub release and reports one
 * only when it is actually newer than the installed build.
 *
 * Two lookup paths, same as the PC app: api.github.com first (gives the real asset URL + notes), then
 * the plain github.com/…/releases/latest redirect as a fallback — the API host is frequently blocked
 * in RU while the normal site still resolves.
 */
object UpdateChecker {
    private const val OWNER = "dutow20162007-create"
    private const val REPO = "DashConnect"

    const val RELEASES_PAGE = "https://github.com/$OWNER/$REPO/releases/latest"

    suspend fun check(currentVersion: String): UpdateInfo? = withContext(Dispatchers.IO) {
        val latest = viaApi() ?: viaRedirect() ?: return@withContext null
        if (isNewer(latest.version, currentVersion)) latest else null
    }

    private fun viaApi(): UpdateInfo? = try {
        val body = httpGet("https://api.github.com/repos/$OWNER/$REPO/releases/latest")
        val o = JSONObject(body)
        val tag = o.optString("tag_name").trimStart('v', 'V')
        var apk = ""
        var sha = ""
        o.optJSONArray("assets")?.let { arr ->
            for (i in 0 until arr.length()) {
                val a = arr.getJSONObject(i)
                val name = a.optString("name")
                if (name.endsWith(".apk.sha256", ignoreCase = true)) sha = a.optString("browser_download_url")
                else if (name.endsWith(".apk", ignoreCase = true)) apk = a.optString("browser_download_url")
            }
        }
        if (tag.isEmpty()) null
        else UpdateInfo(tag, apk.ifEmpty { apkUrlFor(tag) }, o.optString("body"), sha.ifEmpty { null })
    } catch (e: Exception) {
        null
    }

    private fun viaRedirect(): UpdateInfo? = try {
        val c = (URL(RELEASES_PAGE).openConnection() as HttpURLConnection).apply {
            instanceFollowRedirects = false
            connectTimeout = 6000; readTimeout = 6000
            setRequestProperty("User-Agent", "DashConnect")
        }
        val location = c.getHeaderField("Location") ?: ""
        c.disconnect()
        val tag = location.substringAfterLast("/tag/", "").trimStart('v', 'V')
        if (tag.isEmpty()) null else UpdateInfo(tag, apkUrlFor(tag), "", apkUrlFor(tag) + ".sha256")
    } catch (e: Exception) {
        null
    }

    /** Release assets are published as DashConnect-<version>.apk. */
    private fun apkUrlFor(v: String) =
        "https://github.com/$OWNER/$REPO/releases/download/v$v/DashConnect-$v.apk"

    /**
     * Downloads the APK into cache and returns the file, or null on failure. When [sha256Url] is
     * given and the release published a checksum, the file is verified against it and a confirmed
     * MISMATCH fails the download (returns null) so a tampered/corrupt APK is never installed.
     */
    suspend fun download(ctx: Context, url: String, sha256Url: String? = null, onProgress: (Int) -> Unit = {}): File? =
        withContext(Dispatchers.IO) {
            try {
                val dir = File(ctx.cacheDir, "update").apply { mkdirs() }
                val out = File(dir, "DashConnect-update.apk")
                if (out.exists()) out.delete()
                val c = (URL(url).openConnection() as HttpURLConnection).apply {
                    instanceFollowRedirects = true
                    connectTimeout = 15000; readTimeout = 30000
                    setRequestProperty("User-Agent", "DashConnect")
                }
                if (c.responseCode !in 200..299) { c.disconnect(); return@withContext null }
                val total = c.contentLength.toLong()
                var read = 0L
                c.inputStream.use { input ->
                    out.outputStream().use { fout ->
                        val buf = ByteArray(64 * 1024)
                        while (true) {
                            val n = input.read(buf)
                            if (n < 0) break
                            fout.write(buf, 0, n)
                            read += n
                            if (total > 0) onProgress(((read * 100) / total).toInt())
                        }
                    }
                }
                c.disconnect()
                if (out.length() <= 0) return@withContext null
                if (!verifyChecksum(out, sha256Url)) { out.delete(); return@withContext null }
                out
            } catch (e: Exception) {
                null
            }
        }

    /** Fail-open: if the checksum can't be fetched/parsed (older release), returns true; only a real
     *  mismatch returns false. */
    private fun verifyChecksum(file: File, sha256Url: String?): Boolean {
        if (sha256Url.isNullOrEmpty()) return true
        val expected = try {
            val c = (URL(sha256Url).openConnection() as HttpURLConnection).apply {
                instanceFollowRedirects = true; connectTimeout = 10000; readTimeout = 15000
                setRequestProperty("User-Agent", "DashConnect")
            }
            if (c.responseCode !in 200..299) { c.disconnect(); return true }
            val text = c.inputStream.bufferedReader().use { it.readText() }
            c.disconnect()
            Regex("[0-9a-fA-F]{64}").find(text)?.value?.lowercase() ?: return true
        } catch (e: Exception) { return true }

        val md = MessageDigest.getInstance("SHA-256")
        file.inputStream().use { input ->
            val buf = ByteArray(64 * 1024)
            while (true) { val n = input.read(buf); if (n < 0) break; md.update(buf, 0, n) }
        }
        val actual = md.digest().joinToString("") { "%02x".format(it) }
        return actual == expected
    }

    private fun httpGet(url: String): String {
        val c = (URL(url).openConnection() as HttpURLConnection).apply {
            connectTimeout = 6000; readTimeout = 6000
            setRequestProperty("User-Agent", "DashConnect")
            setRequestProperty("Accept", "application/vnd.github+json")
        }
        return try { c.inputStream.bufferedReader().readText() } finally { c.disconnect() }
    }

    /** Numeric, segment-wise compare so 1.1.10 correctly beats 1.1.9. */
    fun isNewer(latest: String, current: String): Boolean {
        val a = latest.split('.', '-').mapNotNull { it.toIntOrNull() }
        val b = current.split('.', '-').mapNotNull { it.toIntOrNull() }
        for (i in 0 until maxOf(a.size, b.size)) {
            val x = a.getOrElse(i) { 0 }
            val y = b.getOrElse(i) { 0 }
            if (x != y) return x > y
        }
        return false
    }
}
